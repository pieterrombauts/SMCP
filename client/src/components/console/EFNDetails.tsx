import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import 'bootstrap/dist/css/bootstrap.min.css';
import { statusReport } from './customTypes'
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import socket from '../Socket';

interface AppProps {
  className?: string;
  reports: statusReport[];
  efnID: string;
  lobbyID: string;
  closeFunction: () => void;
};

const UEFNDetails: React.FC<AppProps> = ( props ) => {
  const commentRef = React.useRef<any>(null);
  const [ currEFN, setCurrEFN ] = useState<statusReport>({_ID: "", _sender: "", _subject: "", _content: "", _status: "open", _time: "", _comments: [] })
  useEffect(() => {
    if (props.reports.length !== 0) {
      var selected_efn = props.reports.filter((report) => { return report._ID === props.efnID });
      if (selected_efn.length === 1) {
        setCurrEFN(selected_efn[0]);
      }
    }
  }, [props.efnID, props.reports])
  function addEFNComment(e: React.MouseEvent<HTMLButtonElement>) {
    socket.emit("ADD_EFN_COMMENT", commentRef.current.value, props.efnID, props.lobbyID);
    commentRef.current.value = "";
  }
  return(
    <div className={props.className}>
      { currEFN !== null && 
        <div key={props.efnID}>
          <p className="efn-id"><small>{currEFN._ID}</small></p>
          <h3 className="efn-header">{currEFN._sender.toUpperCase() + " - " + currEFN._subject}</h3>
          <p className="efn-status">{currEFN._status.toUpperCase()}</p>
          <p className="efn-content">{currEFN._content}</p>
          <hr />
          {currEFN._comments.map((comment) => {
            return(<><p>{comment}</p><hr /></>)
          })}
        </div>
      }
      <Form.Control as="textarea" name="content" rows="5" ref={commentRef} />
      <Button variant="primary" onClick={addEFNComment}>Add Comment</Button>
    </div>
  );
}

export const EFNDetails = styled(UEFNDetails)`
  .efn-id {
    margin-bottom: 3px;
  }
  .efn-header {
    display: inline;
  }
  .efn-status {
    display: inline;
    margin-left: 10px;
  }
  .efn-content {
    margin-top: 10px;
  }
  button {
    margin-top: 10px;
  }
`;