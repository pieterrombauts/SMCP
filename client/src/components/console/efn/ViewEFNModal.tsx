import React from 'react';
import styled from 'styled-components';
import 'bootstrap/dist/css/bootstrap.min.css';
import { EFNDetails } from './EFNDetails'
import { statusReport } from '../customTypes';

interface AppProps {
  className?: string;
  show: boolean;
  closeFunction: () => void;
  reports: statusReport[];
  selectedEfn: string;
  lobbyID: string;
};

export const UViewEFNModal: React.FC<AppProps> = ( props ) => {
  function handleOverlayClick(event: React.MouseEvent<HTMLDivElement>) {
    if (event.target === event.currentTarget) {
      props.closeFunction();
    }
  }
  return(
    <div id="view-efn-modal" className={props.className}>
      <div id="modal-overlay-view-efn" onMouseUp={handleOverlayClick}>
        <div id="modal-content-view-efn">
          <EFNDetails lobbyID={props.lobbyID} reports={props.reports} efnID={props.selectedEfn} closeFunction={props.closeFunction}/>
        </div>
      </div>
    </div>
  );
}

export const ViewEFNModal = styled(UViewEFNModal)`
  visibility: hidden;
  #modal-overlay-view-efn {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9;
    background-color: rgba(0,0,0,0.5);
  }

  #modal-content-view-efn {
    background-color: #FFF;
    position: absolute;
    width: 40%;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    border-radius: 10px;
    padding: 40px;
    z-index: 10;
  }
`;