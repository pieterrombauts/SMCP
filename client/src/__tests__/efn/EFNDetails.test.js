import React from "react";
import 'jest-styled-components'
import toJson from 'enzyme-to-json'
import { EFNDetails } from '../../components/console/efn/EFNDetails'
import { shallow, mount, render} from 'enzyme';

const testReports = [
  {_ID: "EFN0001", _sender: "spartan", _subject: "Test 1", _content: "This is the first test", _status: "open", _time: "2019-11-19T15:00:00", _comments: ["This is a test comment", "This is a second test comment"] },
  {_ID: "EFN0002", _sender: "cronus", _subject: "Test 2", _content: "This is the second test", _status: "open", _time: "2019-11-19T15:00:02", _comments: [] },
  {_ID: "EFN0003", _sender: "ethos", _subject: "Test 3", _content: "This is the third test", _status: "open", _time: "2019-11-19T15:00:05", _comments: [] }
]

describe("EFN details modal", () => {
  it('renders correctly', () => {
    const wrapper = shallow(<EFNDetails reports={testReports} efnID={'EFN0001'} lobbyID={'111111'} closeFunction={jest.fn()}/>)
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  })

  it('displays the right EFN details', () => {
    const wrapper = mount(<EFNDetails reports={testReports} efnID={'EFN0001'} lobbyID={'111111'} closeFunction={jest.fn()}/>)
    expect(wrapper.find('.efn-id').children('small').text()).toBe('EFN0001');
    expect(wrapper.find('.efn-header').text()).toBe('SPARTAN - Test 1');
    expect(wrapper.find('.efn-status').text()).toBe('OPEN');
    expect(wrapper.find('.efn-content').text()).toBe('This is the first test');
  })

  it('displays the right number of comments', () => {
    const wrapper = mount(<EFNDetails reports={testReports} efnID={'EFN0001'} lobbyID={'111111'} closeFunction={jest.fn()}/>)
    expect(wrapper.find('.efn-comment')).toHaveLength(2);
    expect(wrapper.find('.efn-comment').first().text()).toBe('This is a test comment');
    expect(wrapper.find('.efn-comment').last().text()).toBe('This is a second test comment');
  })

  it('renders the comment text box and button', () => {
    const wrapper = mount(<EFNDetails reports={testReports} efnID={'EFN0001'} lobbyID={'111111'} closeFunction={jest.fn()}/>)
    expect(wrapper.find('textarea')).toHaveLength(1);
    expect(wrapper.find('.efn-add-comment-btn').last().text()).toBe("Add Comment");
  })
})