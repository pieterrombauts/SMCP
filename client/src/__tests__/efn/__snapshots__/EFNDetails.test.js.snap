// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EFN details modal renders correctly 1`] = `
.c0 .efn-id {
  margin-bottom: 3px;
}

.c0 .efn-header {
  display: inline;
}

.c0 .efn-status {
  display: inline;
  margin-left: 10px;
}

.c0 .efn-content {
  margin-top: 10px;
}

.c0 button {
  margin-top: 10px;
}

<UEFNDetails
  className="c0"
  closeFunction={[MockFunction]}
  efnID="EFN0001"
  lobbyID="111111"
  reports={
    Array [
      Object {
        "_ID": "EFN0001",
        "_comments": Array [
          "This is a test comment",
          "This is a second test comment",
        ],
        "_content": "This is the first test",
        "_sender": "spartan",
        "_status": "open",
        "_subject": "Test 1",
        "_time": "2019-11-19T15:00:00",
      },
      Object {
        "_ID": "EFN0002",
        "_comments": Array [],
        "_content": "This is the second test",
        "_sender": "cronus",
        "_status": "open",
        "_subject": "Test 2",
        "_time": "2019-11-19T15:00:02",
      },
      Object {
        "_ID": "EFN0003",
        "_comments": Array [],
        "_content": "This is the third test",
        "_sender": "ethos",
        "_status": "open",
        "_subject": "Test 3",
        "_time": "2019-11-19T15:00:05",
      },
    ]
  }
/>
`;
